src/Middleware/ConcurrencyLimiter/src/QueuePolicies/LIFOQueuePolicy.cs

```diff
@@ -42,12 +48,12 @@ public Task<bool> TryEnterAsync()
                 if (_queueLength == _maxQueueCapacity)
                 {
                     _hasReachedCapacity = true;
-                    _buffer[_head].SetResult(false);
+                    _buffer[_head].Complete(false);
                     _queueLength--;
                 }
 
-                // enqueue request with a tcs
-                var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                var tcs = Interlocked.Exchange(ref _cachedResettableTCS, null) ?? new ResettableBooleanCompletionSource(this);
```

Comment:
Shouldn't need Interlocked as its in a lock?
```suggestion
                var tcs = _cachedResettableTCS ??= new ResettableBooleanCompletionSource(this);
```

Reply:
The pooled TCSs are being placed back in that field by a `Volatile.Write` that's not in any lock, and so this change introduces a race condition and breaks unit tests.

I will go run benchmarks on several combinations of locks, and see what's best. Thinking

- Current strategy
- Interlocked for the write, this change for the read
- Locking on the same lock for the write (probably slowest)

Reply:
Ah didn't see was different class returning it. Probably good as now.

Does mean only one gets reused rather than a pool of them?

Reply:
> Does mean only one gets reused rather than a pool of them?

Correct. Generally the call to GetResult() happens inline and returns the ResettableBooleanCompletionSource to the field before the next attempt to retrieve/alloc another.

It's important that we await something like [ThreadPoolAwaitable](https://github.com/aspnet/AspNetCore/blob/caa910ceeba5f2b2c02c47a23ead0ca31caea6f0/src/Servers/Kestrel/Core/src/Internal/Infrastructure/ThreadPoolAwaitable.cs) before calling _next(context) in ConcurrencyLimiterMiddleware so we do not:

A) Run app code (i.e. `_next(context)`) with the LIFOQueuePolicy _bufferLock acquired.
B) Stack dive by calling repeatedly `_next(context)` inline inside LIFOQueuePolicy.OnExit() which could ultimately lead to a StackOverflow

It would be safer to dispatch at the LIFOQueuePolicy layer instead of the ConcurrencyLimiterMiddleware, since the inline-scheduled ValueTask is exposed to end users if they resolve the LIFOQueuePolicy as an IQueuePolicy from DI.

The problem is without inline scheduling, you lose the nice behavior where GetResult() happens inline and returns the ResettableBooleanCompletionSource to the field before the next attempt to retrieve/alloc another.

Reply:
The way the pooling works, is that the queue is full of these awaitables. When one gets awaited on, and has `.GetResult` called, it resets itself and puts itself back into the `_cache` field. So there's a pool, containing: the awaitables in active use + one additional cached awaitable. 

If the queue goes back down, from say 1000 items to 0, then 999 of those thousand will get GCd and 1 will be kept. What this optimizes for is the case of heavy load / ddos, where the queue is stuck at full and requests keep coming in; it avoids allocating extra completion sources in this state.

-------------------
So I found why the unit tests broke, and it was because the queue policy needs to reset the cache to null to prevent to awaitable from getting regrabbed before it's ready.

Ended up testing five different scenarios, by far the best performance was the simple approach of:

```
// no locking or checking within the CompletionSource reset
_mrvts.Reset();
_queue._cachedResettableTCS = this;
```

```
// no additional locking within the queue policy either, setting to null after
var tcs = _cachedResettableTCS ??= new ResettableBooleanCompletionSource(this);
_cachedResettableTCS = null;
```

This feels sloppy but safe. The policy can never double grab from the cache, because it's stuck in a lock, and if the reset fails to go through (which happens <1% on my benchmarks) worst case is allocating a couple hundred bytes.